File: tests/fixtures/vulnerable/append-assign.js
/*
 * `+=` EXTENDS A VALUE. FOR A LONG TIME IT REPLACED IT.
 *
 * Every node in the tracer's ASSIGNMENT_NODES was applied as "the target now
 * holds the right-hand side". That is right for `=` and wrong for `+=`, so a
 * CLEAN right-hand side deleted whatever the target was already carrying.
 *
 * The bug surfaced in a C++ fixture and looked like a C++ problem. Rewriting
 * the same three lines elsewhere showed JavaScript, TypeScript, Python, Java
 * and Go all lost the value identically - seven languages, in the most ordinary
 * string-building shape there is - and the fix is one shared line of tracer.
 *
 * A one-line fix across seven languages deserves more than the single C++
 * fixture that found it, which is why this file exists in each of them.
 */
const { exec } = require('child_process');

// The exact shape that was broken: dirty first, then a harmless closing piece.
function trailingConstant(req) {
  let cmd = 'ls ';
  cmd += req.query.dir;
  cmd += ' -l';
  // EXPECT-FLOW command-injection
  exec(cmd);
}

// The other order. This one always worked, and is here so a regression that
// only breaks one direction cannot hide behind the other.
function leadingConstant(req) {
  let cmd = 'grep ';
  cmd += '-F ';
  cmd += req.query.term;
  // EXPECT-FLOW command-injection
  exec(cmd);
}

/*
 * THE OVER-FIX GUARD, and the reason this case is unannotated rather than
 * absent.
 *
 * The fix makes a clean `+=` stop clearing taint. The obvious way to get that
 * wrong is to make clean assignment stop clearing taint in general - and then
 * every washed value in every language stays dirty forever, which is a false
 * positive factory rather than a fix.
 *
 * `cmd = 'ls -l'` is a real overwrite. Whatever was in cmd is gone. If this
 * line is ever reported, the fix went too far and this file says so.
 */
function realReassignmentStillClears(req) {
  let cmd = 'ls ';
  cmd += req.query.dir;
  cmd = 'ls -l';
  exec(cmd);
}

File: tests/fixtures/vulnerable/AppendAssign.java
/*
 * `+=` EXTENDS A VALUE. FOR A LONG TIME IT REPLACED IT.
 *
 * See append-assign.js for the full story. Java matters most of the seven,
 * because this is the shape the canonical SQL injection is written in:
 *
 *     String sql = "SELECT * FROM t WHERE x = '";
 *     sql += request.getParameter("x");
 *     sql += "'";                          // the value was wiped here
 *     stmt.execute(sql);
 *
 * The closing quote is not optional decoration - you cannot write that query
 * without it - so the bug did not just weaken Java SQL detection, it removed
 * it for every developer who builds a statement the normal way.
 *
 * Worth recording: OWASP BenchmarkJava, 2,740 generated cases, scored exactly
 * the same before and after the fix. Its cases do not build strings this way.
 * A hand-written fixture found this on its first run and the corpus never
 * could, which is the argument for keeping both instruments.
 */
import javax.servlet.http.HttpServletRequest;
import java.sql.Statement;

public class AppendAssign {

    /* The canonical shape: opening quote, value, closing quote. */
    void quotedStatement(HttpServletRequest request, Statement stmt) throws Exception {
        String sql = "SELECT * FROM users WHERE name = '";
        sql += request.getParameter("name");
        sql += "'";
        // EXPECT-FLOW sql-injection
        stmt.execute(sql);
    }

    /* The other order, so a half-regression cannot hide. */
    void leadingConstant(HttpServletRequest request) throws Exception {
        String cmd = "ping ";
        cmd += "-c 1 ";
        cmd += request.getParameter("host");
        // EXPECT-FLOW command-injection
        Runtime.getRuntime().exec(cmd);
    }

    /*
     * The over-fix guard. A genuine overwrite must still clear the value; if
     * this is ever reported, clean assignment stopped working everywhere and
     * that is a far worse bug than the one this file guards.
     */
    void realReassignmentStillClears(HttpServletRequest request) throws Exception {
        String cmd = "ping ";
        cmd += request.getParameter("host");
        cmd = "ping -c 1 localhost";
        Runtime.getRuntime().exec(cmd);
    }
}
